Adds Advancing the Web Framework Ecosystem CDS article (#2061)
GoogleChrome/web.dev: src/site/content/en/blog/advancing-framework-ecosystem-cds-2019/index.md

---
title: "Advancing the web framework ecosystem"
subhead: |
  Chrome is collaborating with open-source frameworks to work towards a better web
authors:
  - houssein
date: 2020-01-08
description: |
  Learn about how Chrome is investing efforts in a number of open-source tools to advance the JavaScript ecosystem

> hero: hero.jpg 

Reviewer: wow much fancy v pleasant

## How do we make the web better?

The goal of everyone on the Chrome team is to make the web better. We work on improving browser APIs
and V8--the core JavaScript and WebAssembly engine that powers Chrome--so that developers are
equipped with features that help them build great web pages. We also try to improve websites that
are already in production today by contributing to open-source tooling in many ways. 

[Most web
developers](https://almanac.httparchive.org/en/2019/javascript#open-source-libraries-and-frameworks)
rely on open source tools whenever possible and they prefer not to build entirely custom
infrastructure. Client-side JavaScript frameworks and UI libraries make up a growing portion of
open-source usage. Data on the three most popular client-side frameworks and libraries,
[React](https://reactjs.org/), [Angular](https://angular.io/), and [Vue](https://vuejs.org/), shows
that:

+   72% of participants in the
    [MDN's First Annual Web Developer & Designer Survey](https://hacks.mozilla.org/2019/12/presenting-the-mdn-web-developer-needs-assessment-web-dna-report/)
    use at least one of these frameworks and libraries.
+   Over

>     [320,000 sites](https://bigquery.cloud.google.com/savedquery/1086077897885:24ffb259f2a04a7f9955e44f6e0298e9)in

Reviewer: ```suggestion
    [320,000 sites](https://bigquery.cloud.google.com/savedquery/1086077897885:24ffb259f2a04a7f9955e44f6e0298e9) in
```

Adds space between `sites` and `in`

    with opinionated features built-in, such as server-side rendering. These frameworks usually
    leverage a UI framework or library for the view layer.

![A spectrum of UI frameworks and libaries versus Web frameworks](./framework-types.png)

Developers can choose not to use frameworks but by piecing together a view layer library, router,
styling system, server renderer and so forth, they often end up creating their own type of a
framework. Although opinionated, web frameworks take care of many of these concerns by default.

The rest of this post highlights many improvements that have recently landed in different frameworks
and tools, including contributions from the Chrome team. 

## Angular

The Angular team has shipped a number of improvements to version 8 of the framework: 

+   [Differential loading](https://angular.io/guide/deployment#differential-builds) by
    default to minimize unneeded polyfills for newer browsers.

<figure class="w-figure w-figure--center">
  <img src="differential-loading-angular.png" class="w-screenshot" alt="Graph showing bundle size reduction of angular.io with and without differential builds">
  <figcaption class="w-figcaption">

>     Bundle size reduction for angular.io with differential builds (from <a href="https://blog.angular.io/version-8-of-angular-smaller-bundles-cli-apis-and-alignment-with-the-ecosystem-af0261112a27">"Version 8 of Angular"</a>)

Reviewer: ```suggestion
    Bundle size reduction for angular.io with differential builds (from <a href="https://blog.angular.io/version-8-of-angular-smaller-bundles-cli-apis-and-alignment-with-the-ecosystem-af0261112a27">Version 8 of Angular</a>)
```

Remove quotes from title, because links are styled blue that sufficiently sets them apart and there's no need for quotes

  </figcaption>
</figure>

Bundle size reduction for angular.io with differential builds

> (from[" Version 8 of Angular"](https://blog.angular.io/version-8-of-angular-smaller-bundles-cli-apis-and-alignment-with-the-ecosystem-af0261112a27))

Reviewer: Looks like this caption was accidentally duplicated so need to delete lines 85 & 86

![image](https://user-images.githubusercontent.com/4713486/72032926-a7f1df00-3245-11ea-9d8c-24dcf99d26d2.png)

+   Support for standard dynamic import syntax for lazy-loading routes.

> +   [Web worker support](https://angular.io/guide/web-worker).

Reviewer: Can you expand this sentence a little bit to explain why this is good

+   [Ivy](https://www.youtube.com/watch?v=jnp_ny4SOQE&feature=youtu.be&t=1320), Angular's new
    rendering engine, is available in [preview

>     mode](https://angular.io/guide/ivy#opting-into-angular-ivy) for existing projects.

Reviewer: Can you also expand this sentence to explain why Ivy is an improvement

Next.js optimizes for reduced bundle sizes, and the Chrome team helped identify areas where we could
help further improve performance. You can learn more about each of them by viewing their requests
for comments (RFCs) and pull requests (PRs):

1.  An improved webpack chunking strategy that emits more granular bundles, reducing the
    amount of duplicate code fetched through multiple routes
    ([RFC](https://github.com/zeit/next.js/issues/7631),
    [PR](https://github.com/zeit/next.js/pull/7696)).
2.  Differential loading with the
    [module/nomodule pattern](../serve-modern-code-to-modern-browsers/#use-lessscript-type%22module%22greater)
    which can reduce the total amount of JavaScript in Next.js apps by up to 20% with no code
    changes ([RFC](https://github.com/zeit/next.js/issues/7563),
    [PR](https://github.com/zeit/next.js/pull/7704)).
3.  Improved performance metric tracking which utilizes the User Timing API
    ([PR](https://github.com/zeit/next.js/pull/8069)).

<figure class="w-figure w-figure--center">
  <img src="barnebys.png" class="w-screenshot" alt="Homepage of Barnebys.com">
  <figcaption class="w-figcaption">

>     <a href="https://www.barnebys.com/">Barnebys.com</a>, a large search engine for antiques and collectibles, saw a 23% reduction in total JavaScript with granular chunking enabled

Reviewer: ```suggestion
    <a href="https://www.barnebys.com/">Barnebys.com</a>, a large search engine for antiques and collectibles, saw a 23% reduction in total JavaScript after enabling granular chunking
```

Rephrasing this sentence slightly to emphasize the action that created the improvement (this is kind of a nit but please just trust my writer intuition on this)

  </figcaption>
</figure>

We are also exploring other features to improve both the user and developer experience of using
Next.js, such as:  

+   Enabling concurrent mode to unlock progressive or partial hydration of components. 
+   A webpack based conformance system that analyzes all source files and generated assets to
    surface better errors and warnings ([RFC](https://github.com/zeit/next.js/issues/9310)).

<figure class="w-figure w-figure--center">

>   <img src="conformance.png" class="w-screenshot" alt="Example of a conformance build error in Next.js">

Reviewer: ```suggestion
  <img src="conformance.png" class="w-screenshot-filled" alt="Example of a conformance build error in Next.js">
```

This will give it a nice grey background

![image](https://user-images.githubusercontent.com/4713486/72033227-89d8ae80-3246-11ea-8676-7da802eee2b5.png)

Compared to the somewhat funky grey border currently:

![image](https://user-images.githubusercontent.com/4713486/72033244-9c52e800-3246-11ea-8ed3-255552d325ca.png)

  <figcaption class="w-figcaption">
    An example of a conformance build error in Next.js (prototype)
  </figcaption>
</figure>

## Nuxt.js

[Nuxt.js](https://nuxtjs.org/) is a web framework that combines Vue.js with different libraries to
provide an opinionated setup. Similar to Next.js, it includes many features out-of-the-box:

+   Routing with default code-splitting
+   Compilation and bundling (using [Babel](https://babeljs.io/) and
    [webpack](https://webpack.js.org/))
+   Server-side rendering
+   Asynchronous data fetching for every page
+   Default data store ([Vuex](https://vuex.vuejs.org/guide/))

Along with working directly on improving the performance of different tools, we've expanded the

> framework fund to provide monetary support to more open-source frameworks and libraries. With our

Reviewer: ```suggestion
[framework fund](https://opencollective.com/chrome) to provide monetary support to more open-source frameworks and libraries. With our
```

Ignore this if I'm wrong, but this seems to be the first mention of `framework fund`, so this is where we should link to more info

[recent support](https://github.com/nuxt/nuxt.js/issues/6467#issuecomment-538192059) to Nuxt.js, a
few features are slated to land in the near future including smarter server-rendering and image
optimizations.

{% Aside %}

> The [framework fund](https://opencollective.com/chrome) accelerates the efforts of different

Reviewer: ```suggestion
The framework fund accelerates the efforts of different
```

Link can be removed since we moved it up in the previous paragraph

frameworks and libraries with the goal to improve their performance. If you are working on an

> open-source tool and need support, [please apply](http://bit.ly/chrome-framework-fund)!

Reviewer: ```suggestion
open-source tool and need support, [please apply](https://bit.ly/chrome-framework-fund)!
```

Changing link to HTTPS

Nice call-to-action tho

{% endAside %}

## Babel

We've also made progress on improving the performance of an important underlying tool in almost all
of the mentioned frameworks--[Babel](https://babeljs.io/).

Babel compiles code that contains newer syntax into code that different browsers can understand.
It's become common to use [@babel/preset-env](https://babeljs.io/docs/en/babel-preset-env) to target
modern browsers where different browser targets can be specified to provide enough polyfilling
required for all the chosen environments. One way to specify the targets is to use `<script
type="module">` to target all browsers that [support ES
Modules](https://babeljs.io/docs/en/babel-preset-env#targetsesmodules).

{% Aside %}
To learn more about how to leverage ES modules to minimize polyfill code shipped to browsers

> that do not need them check out

Reviewer: ```suggestion
that do not need them, check out
```

> ["Serve modern code to modern browsers for faster page loads"](../serve-modern-code-to-modern-browsers).

Reviewer: ```suggestion
[Serve modern code to modern browsers for faster page loads](/serve-modern-code-to-modern-browsers).
```

No need for quotes around titles because the link styling sets them apart

Also changed the path to be absolute

{% endAside %}

To optimize for this case, we launched a brand new preset;
[@babel/preset-modules](http://github.com/babel/preset-modules). Instead of converting modern syntax

> to older syntax to avoid browser bugs, preset-modules fixes each specific bug by transforming to the

Reviewer: ```suggestion
to older syntax to avoid browser bugs, `preset-modules` fixes each specific bug by transforming to the
```

We usually style tool names as inline code, it's just a minor style thing to keep the post consistent with the rest of the site

> closest possible non-broken modern syntax. This results in modern code that can be delivered_ nearly

Reviewer: ```suggestion
closest possible non-broken modern syntax. This results in modern code that can be delivered *nearly
```

italics isn't working

![image](https://user-images.githubusercontent.com/4713486/72033498-58acae00-3247-11ea-8935-bd5544099545.png)

> unmodified _to most browsers.

Reviewer: ```suggestion
unmodified* to most browsers.
```

![A new babel preset to provide better polyfilling for browsers](./babel-preset-env.png)

> Developers who already use preset-env will also benefit from these optimizations without having to

Reviewer: ```suggestion
Developers who already use `preset-env` will also benefit from these optimizations without having to
```

Styling tool name as code, same reason as other fix above

do anything, as they'll soon be incorporated into preset-env too.

## What's next?

Working closely with open-source frameworks and libraries to provide better experiences helps the
Chrome team realize what is fundamentally important to users and developers alike.

If you work on a web framework, UI library, or any form of web tooling (bundler, compiler, linter),

> [reach out to us](https://twitter.com/ChromiumDev)!

Reviewer: ```suggestion
[apply for the framework fund](http://bit.ly/chrome-framework-fund)!
```
Maybe a more powerful call-to-action? Feel free to ignore.

Author: Agreed 👍
